conference/LinuxPlumbers2021/xdp-multi-buff.org

```diff
@@ -226,6 +226,31 @@ out:
       return XDP_PASS;
 }
 #+end_src
+
+** Slide: XDP multi-buffers: new eBPF helpers (5/5)			   :export:
+- /bpf_xdp_output/ and /bpf_perf_event_output/:
+  - Helpers are updated and can copy the content of all buffers
+#+begin_src C
+static unsigned long bpf_xdp_copy(void *dst_buff, const void *ctx,
+                                  unsigned long off, unsigned long len)
+{
+        if (likely(!xdp_buff_is_mb(xdp))) {
+                memcpy(dst_buff, xdp->data + off, len);
+                return 0;
+        }
+        ...
+        /* If we need to copy data from the base buffer do it */
+        if (off < base_len) {
+                ...
+        }
+        /* Copy any remaining data from the fragments */
+        for (i = 0; len && i < sinfo->nr_frags; i++) {
+                ...
+        }
+      return 0;
```

Comment:
can you please fix the code indentation?

Reply:
Done!